Implemented PMI masking for T5-Pretraining.

OfekGlick opened the pull request:
Changes include adding the PMI vocabulary, n-grams range from size 2-5, Implementing PMI masking in the Data Collator,
Adding relevant flags in the flag file, and unit testing for the PMI masking methods.

OfekGlick committed:
Implemented PMI masking for T5-Pretraining.
Changes include adding the PMI vocabulary, n-grams range from size 2-5,
Implementing PMI masking in the Data Collator,
Adding relevant flags in the flag file, and unit testing for the PMI masking methods.

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
     return shifted_input_ids
 
 
+def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, segments_to_merge, ngrams_vocab_set):
```
Missing docstring and type annotations.

zbambergerNLP replied:
Add a TODO: vectorize this function with pytorch/numpy

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
+        if segment in ngrams_vocab_set:
+            possible_merges.append(list(range(offset + i, offset + i + n_grams)))
+
+    random.shuffle(possible_merges)
```
nit: Are you sure that given random.setseed() performed earlier , that this call will be deterministic?

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
+        segments_to_merge.append(seg_inds)
+
+
+def pmi_word_mask(input_tokens, pmi_vocab, max_predictions=512, mlm_probability=0.5):
```
TODO: vectorize this function with pytorch or numpy

zbambergerNLP replied:
Missing type annotations

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
+    whole_words_lists = [[]]
+    whole_words = whole_words_lists[0]
+    for (i, token) in enumerate(input_tokens):
+        if token == '</s>' or token == '<pad>':
```
'\<\/s\>' should be a constant and same with '\<pad\>'. Avoid comparing against naked strings.

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
+
+def pmi_word_mask(input_tokens, pmi_vocab, max_predictions=512, mlm_probability=0.5):
+    """
+    Create a mask for PMI-based corruption for a sample.
```
Add another line or two which describe the high-level logic behind creating the masks.

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
+        if len(whole_words) == 0:
+            continue
+        added_offset = len(whole_words)
+        whole_words = ' '.join(whole_words)
```
What about whole words that are separated by linebreaks or other forms of spacing?

OfekGlick replied:
Can you expand on this point?
From analyzing the way the T5 tokenizing works, most words have their own specific token. Where every token (other then the first word) will have a "_" before the word itself. 
If a word is comprised of a number of tokens, the first token will have a "_" before it, while the rest of the tokens will not. So I believe these "if" statemesnts cover 99% of cases.

zbambergerNLP replied:
You make a good point about sentence-piece tokens starting with an underscore. I think we can resolve this comment. Let's exercise caution when we perform tokenization (break down words into tokens) on the basis of spaces, since we'd also need to account for other forms of spaces such as tabs and linebreaks. If the logic is determined by the common underscore that precedes words in sentence-piece, then things are good by my standards.

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
+
+    # whole_words_indexes is a list of lists of ints, each list is the
+    # indices part of the whole word to be masked, i.e. the segment to be considered for masking
+    random.shuffle(candidates)
```
Here too we should be clear that this is can be deterministic (by setting a seed earlier in the code)

zbambergerNLP commented on corruption_t5_test.py:
Need new and robust tests for each of the new functions.

OfekGlick replied:
Can you expand on this point?

zbambergerNLP replied:
The unit tests you created in pmi_corruption_test.py should be here. Each new function you created in corruption_t5.py should be tested in this file in particular. You should identify edge cases and test against them as well (e.g., what happens if you encounter the empty string, vs a 1-line text, vs. a multi-line text with different spacing types).

zbambergerNLP commented on fine_tune_t5.py:
```diff
@@ -20,6 +20,8 @@
     parser = HfArgumentParser((flags.ModelArguments, flags.DataTrainingArguments, flags.TrainingArguments))
     model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+
+    os.environ["WANDB_API_KEY"] = '0bc4005546f2e4d5791ff50a0eeb476ee687d5c6'
```
Remove from code. This is a public repo.

zbambergerNLP commented on flags.py:
```diff
@@ -102,7 +102,7 @@ class TrainingArguments:
     )
     deepspeed: bool = field(
-        default=True, metadata={"help": "Whether to use deepspeed for training."}
+        default=False, metadata={"help": "Whether to use deepspeed for training."}
```
Change back to True

zbambergerNLP commented on flags.py:
```diff
@@ -241,11 +245,11 @@ class DataTrainingArguments:
     )
     pre_training_dataset_paths: Optional[str] = field(
-        default="wikipedia,bookcorpus",
+        default="bookcorpus",
```
Change this back to include wikipedia as well

zbambergerNLP commented on flags.py:
```diff
@@ -210,7 +214,7 @@ class DataTrainingArguments:
     )
     percent_of_dataset: Optional[int] = field(
-        default=100,
+        default=2,
```
Change this back to 100 percent.

zbambergerNLP commented on flags.py:
```diff
@@ -241,11 +245,11 @@ class DataTrainingArguments:
     )
     pre_training_dataset_names: Optional[str] = field(
-        default="20220301.en,",  # Only wikipedia has a name, bookcorpus is just a path.
+        default="",  # Only wikipedia has a name, bookcorpus is just a path.
```
Change back

zbambergerNLP commented on pmi_corruption_test.py:
Merge with existing t5_corruption tests.

zbambergerNLP commented on pmi_corruption_test.py:
```diff
@@ -0,0 +1,169 @@
+        random.seed(42)
+
+    @parameterized.named_parameters(
+        *[{
```
Do not use this notation, you should have a defined dictionary here with clear testcase names, and potential modifications to parameters that are critical to the function under test.

zbambergerNLP replied:
See the examples in the existing test file, corruption_t5_test.py

zbambergerNLP commented on pmi_corruption_test.py:
```diff
@@ -0,0 +1,169 @@
+            mlm_probability,)
+        self.assertIsNotNone(mask_labels_for_sample)
+
+    @parameterized.named_parameters(
+        {
+            TESTCASE_NAME: f"test_pmi_mask_word",
+            EXAMPLES: DEMO_TEXTS,
+            PMI_VOCAB: PMI_DEMO_VOCAB,
+        },
+    )
```
Make a unit test (i.e., a named parameter dictionary) for each of the tests above. It should be clear what is happening in each unit test.

OfekGlick replied:
Can you expand on this point as well? what is not clear about what is happening?

zbambergerNLP replied:
See an example of the format I am looking for: 

```
@parameterized.named_parameters(
        {
            'testcase_name': 'short post',
            'input_argument': 'Israel is a terrorist state.',
            'is_pro_israel': False,
        },
        {
            'testcase_name': 'another short post',
            'input_argument': 'Israel is an apartheid state.',
            'is_pro_israel': False,
        },
        {
            'testcase_name': 'medium length post',
            'input_argument': 'Hamas was justified. What else can Palestinians do in the face of occupation?',
            'is_pro_israel': False,
        },
        {
            'testcase_name': 'short supportive',
            'input_argument': 'In conflicts, civilian populations often suffer the most, highlighting the need for '
                              'peaceful resolution and humanitarian aid.',
            'is_pro_israel': True,
        },
        {
            'testcase_name': 'medium supportive',
            'input_argument': 'The historical and ongoing conflict in the Middle East involves complex geopolitical '
                              'and social factors, making simplistic labels misleading.',
            'is_pro_israel': True,
        },
        {
            'testcase_name': 'long supportive',
            'input_argument': 'It is crucial to consider the perspectives of international law and human rights '
                              'organizations in understanding the dynamics of the Israel-Palestine conflict.',
            'is_pro_israel': True,
        }
    )
    def test_generate_response(
            self,
            input_argument,
            is_pro_israel,
            audience=src_persuade.Audience.MODERATE,
            counterargument_length=src_persuade.Length.FEW_SENTENCES,

    ):
```
    
    The point of creating a dictionary for each unit test is that it's very easy to then pinpoint which test went wrong, and how that test was configured.

zbambergerNLP commented on pre_train_t5.py:
```diff
@@ -29,11 +29,14 @@ def main():
     accelerator.print('data_args:', data_args)
-
     accelerator.print(f'Started pre-training a {model_args.model_name_or_path} model')
+    os.environ["WANDB_API_KEY"] = '0bc4005546f2e4d5791ff50a0eeb476ee687d5c6'
```
Remove this

zbambergerNLP reviewed (commented):
Please see my comments in this review. Great start!

zbambergerNLP reviewed (commented):
Answered your questions and clarifications

OfekGlick committed:
Added constants for T5 constants such as pad token and space token.
Configured the seed to be deterministic across varied modules and files
Specified various unit tests.
Reverted flags back to their original values.

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -43,7 +43,27 @@ def shift_tokens_right(
+        start_inds: List[int],
+        segments_to_merge: List[List[int]],
+        ngrams_vocab_set: Set[str],
+):
```
Missing return type

OfekGlick replied:
Nothing is to be returned

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -58,32 +78,41 @@ def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, se
+        pmi_vocab: Set[str],
+        max_predictions=512,
+        mlm_probability=0.5,
+):
```
Missing return type

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -43,7 +43,27 @@ def shift_tokens_right(
+        start_inds: The indices of the start of each word in the whole sequence.
+        segments_to_merge: A list of segments to merge.
+        ngrams_vocab_set: The set of ngrams to use for PMI-based corruption.
+    """
```
Missing description of returned object

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -58,32 +78,41 @@ def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, se
         pmi_vocab: The set of ngrams to use for PMI-based corruption.
         max_predictions: The maximum number of tokens to mask.
         mlm_probability: The probability of masking a token.
     """
```
Missing description of returned object

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -500,6 +529,7 @@ def corrupt_for_vanilla_t5(
         pmi: Whether to use PMI-based corruption.
         ngram_vocab_set: The set of ngrams to use for PMI-based corruption.
         tokenizer: The tokenizer to use for PMI-based corruption.
+        seed: The seed to use for random number generation.
```
No need for the seed in this part of the code. Within the test, set the seed during 'setup()', or at the beginning of the unit test. You can even pass in a seed as part of a parameterized unit tests. You should not set the seed or deal with logic surrounding it within the function itself.

OfekGlick replied:
I feel like this should depend on the way we decide to seed.
https://stackoverflow.com/questions/65685060/unique-seed-acrossing-multiple-imported-files-with-random-module-python
According to this answer, in the native `random` module, for random calls to be deterministic across different files, you need to pass the random seed. To avoid that, you can just set it at each file.

To avoid it all together, the `random` module should not be used and should be switched for something else

zbambergerNLP replied:
See CorruptionTest in corrruption_t5_test.py. You'll see there that if you set the seed at the start of the test, you will get deterministic behavior. In those unit tests, we compare corruption functions on specific tensor inputs, and expect to get back specific tensor outputs.

https://github.com/zbambergerNLP/ml_training/blob/master/corruption_t5_test.py#L39-L91

zbambergerNLP commented on corruption_t5.py:
```diff
@@ -515,7 +545,7 @@ def corrupt_for_vanilla_t5(
     #  and not a single tensor of rank 3.
     batch['attention_mask'] = torch.asarray(batch['attention_mask'])
-
+    random.seed(seed)
```
Drop this here, and add it to the appropriate test function

OfekGlick replied:
Why? since I am using `random` for my seeding, unless I do it in this instance the random calls will not be deterministic.

OfekGlick replied:
See my answer above

zbambergerNLP replied:
See CorruptionTest in `corrruption_t5_test.py`. You'll see there that if you set the seed at the start of the test, you will get deterministic behavior. In those unit tests, we compare corruption functions on specific tensor inputs, and expect to get back specific tensor outputs.

zbambergerNLP commented on corruption_t5_test.py:
```diff
@@ -254,6 +360,94 @@ def test_corrupt_for_vanilla_t5(
+            PMI_DEMO_VOCAB,
+            max_predictions,
+            mlm_probability, )
+        self.assertIsNotNone(mask_labels_for_sample)
```
Make more complex assumptions. Since you set the seed for each of the unit tests, you can always expect the same results. Make sure that you keep track of what the result should be, make sure it makes sense to you, and then compare the tensor you get from the function with the one you expect (as passed in via parameterized unit tests).

zbambergerNLP commented on corruption_t5_test.py:
```diff
@@ -254,6 +360,94 @@ def test_corrupt_for_vanilla_t5(
+            pmi_vocab,
+            self.tokenizer,
+        )
+        self.assertIsNotNone(predicted_mask_labels)
```
See my comment above about more robust unit testing.

OfekGlick replied:
This method doesn't have many varied uses. I don't know what other reasonable cases to test

zbambergerNLP reviewed (commented):
Nice progress! The core issue at this point is to make the unit tests more robust.

zbambergerNLP commented:
I think we're nearly there. I responded to your recent comments.

OfekGlick committed:
Moved random-seed from function to unittests.
Implemented more unittests to test specific pmi word masking.
Reverted flags back to their original value

zbambergerNLP reviewed (approved):
Looks great! Awesome work :D
